dbms/src/Interpreters/CrossJoinProbeHelper.cpp

```diff
@@ -388,44 +389,156 @@ Block crossProbeBlockImpl(
+        {
+            src_left_columns[i] = probe_process_info.block.getByPosition(i).column.get();
+        }
+        IColumn::Filter::value_type filter_column_value{};
+        if constexpr (has_null_map)
+        {
+            // todo use column->filter(null_map) to construct the result block in batch
+            for (size_t i = 0; i < probe_process_info.block.rows(); ++i)
+            {
+                if ((*probe_process_info.null_map)[i])
+                {
+                    CrossJoinAdder<KIND, STRICTNESS>::addNotFound(
+                        dst_columns,
+                        num_existing_columns,
+                        src_left_columns,
+                        num_columns_to_add,
+                        i,
+                        filter_ptr,
+                        current_offset,
+                        offset_ptr,
+                        probe_process_info.max_block_size);
+                    if (filter_ptr != nullptr)
+                        filter_column_value = (*filter_ptr)[i];
+                }
+            }
+        }
+        /// construct fill filter and offset column
+        if (!dst_columns[0]->empty())
+        {
+            assert(dst_columns[0]->size() == probe_process_info.row_num_filtered_by_left_condition);
+            if (filter_ptr != nullptr)
+            {
+                for (size_t i = 0; i < probe_process_info.row_num_filtered_by_left_condition; ++i)
+                {
+                    (*filter_ptr)[i] = filter_column_value;
+                }
+            }
+            for (size_t i = 0; i < probe_process_info.row_num_filtered_by_left_condition; ++i)
+            {
+                (*offset_ptr)[i] = i + 1;
+            }
+        }
+        probe_process_info.all_rows_joined_finish = true;
+        return {probe_process_info.result_block_schema.cloneWithColumns(std::move(dst_columns)), false};
+    }
+    assert(probe_process_info.next_right_block_index < right_blocks.size());
+
+    Block right_block = right_blocks[probe_process_info.next_right_block_index];
+    assert(right_block.rows() > 0);
+    size_t right_row = right_block.rows();
+
+    Block block = probe_process_info.result_block_schema.cloneEmpty();
+    for (size_t i = 0; i < num_existing_columns; ++i)
+    {
+        /// left columns
+        assert(block.getByPosition(i).column != nullptr);
+        Field value;
+        probe_process_info.block.getByPosition(i).column->get(probe_process_info.start_row, value);
+        block.getByPosition(i).column = block.getByPosition(i).type->createColumnConst(right_row, value);
+    }
+    for (size_t i = 0; i < right_block.columns(); i++)
+    {
+        /// right columns
+        block.getByPosition(i + num_existing_columns).column = right_block.getByPosition(i).column;
+    }
+    if constexpr (KIND == ASTTableJoin::Kind::Cross_LeftOuterSemi)
+    {
+        /// extra match_helper column for LeftOuterSemi join
+        auto helper_index = block.columns() - 1;
```

Comment:
Better to pass the column index as parameter directly.

Reply:
You mean make `helper_index` as an argument for `crossProbeBlockShallowCopyRightBlockImpl`?

Reply:
Yes, make implicit convention explicit

Reply:
But it is assumed that the `helper` column index is the last column in the block in our current code:
https://github.com/pingcap/tiflash/blob/0d09036d264d253c32f0caaca36a44cb70f64fcd/dbms/src/Interpreters/CrossJoinProbeHelper.cpp#L272-L274

and
https://github.com/pingcap/tiflash/blob/0d09036d264d253c32f0caaca36a44cb70f64fcd/dbms/src/Interpreters/JoinPartition.cpp#L815-L821

Reply:
ok